File: src/leo_flow/dashboard/static/symbolwise-replay.js
"use strict";

(() => {
  const node = (id) => document.getElementById(id);
  if (!node("evidence-symbolwise")) return;
  const parts = window.location.pathname.split("/").filter(Boolean);
  if (parts.length !== 2 || parts[0] !== "recordings") return;
  let recordingId;
  try { recordingId = decodeURIComponent(parts[1]); } catch (_error) { return; }

  const colors = ["#80d8ff", "#fff176", "#ff8a80", "#69f0ae", "#ce93d8"];
  let loaded = false;
  let generation = 0;
  let payload = null;
  let context = null;

  function setState(value, message) {
    node("evidence-symbolwise-state").dataset.state = value;
    node("evidence-symbolwise-state").textContent = message;
    node("evidence-symbolwise-badge").textContent = value === "ready" ? "Candidate evidence" : value;
    node("evidence-symbolwise-badge").dataset.tone = value === "error" ? "error" : "warning";
  }

  async function json(path) {
    const response = await fetch(path, {headers: {accept: "application/json"}, credentials: "same-origin"});
    let body = {};
    try { body = await response.json(); } catch (_error) { /* reported below */ }
    if (!response.ok) {
      const error = new Error(body?.error?.message || `request failed (${response.status})`);
      error.status = response.status;
      throw error;
    }
    return body;
  }

  function checked(name) {
    return [...document.querySelectorAll(`#evidence-controls input[name="${name}"]:checked`)].map((item) => item.value);
  }

  function parameters() {
    const selectedRecordings = new Set(checked("radio"));
    const radios = (context?.recordings || []).filter((item) => selectedRecordings.has(item.recording_id)).map((item) => item.radio_id);
    const query = new URLSearchParams();
    const radioInputs = document.querySelectorAll('#evidence-controls input[name="radio"]');
    const lnbInputs = document.querySelectorAll('#evidence-controls input[name="lnb"]');
    const receiverInputs = document.querySelectorAll('#evidence-controls input[name="receiver"]');
    if ((radioInputs.length && !radios.length) || (lnbInputs.length && !checked("lnb").length) || (receiverInputs.length && !checked("receiver").length)) return null;
    if (radios.length) query.set("radio_ids", [...new Set(radios)].sort().join(","));
    const lnbs = checked("lnb").sort();
    const receivers = checked("receiver").sort();
    if (lnbs.length) query.set("lnb_ids", lnbs.join(","));
    if (receivers.length) query.set("receiver_chain_ids", receivers.join(","));
    return query.toString();
  }

  function selectedPattern(point) {
    const values = new Set(checked("pattern"));
    return point.pattern_role === "qin-exact" ? values.has("qin") : values.has(`surrogate-${point.codebook_index}`);
  }

  function appendCell(row, text, heading = false) {
    const cell = document.createElement(heading ? "th" : "td");
    cell.textContent = String(text);
    if (heading) cell.scope = "row";
    row.append(cell);
  }

  function renderWindow() {
    const body = node("evidence-symbolwise-window-body");
    body.replaceChildren();
    const index = Number(node("evidence-symbolwise-window").value);
    node("evidence-symbolwise-window-output").textContent = String(index);
    for (const stream of payload?.streams || []) {
      const point = stream.windows[index];
      if (!point) continue;
      for (const pattern of point.patterns.filter(selectedPattern)) {
        const row = document.createElement("tr");
        appendCell(row, `${pattern.candidate_label} · ${stream.radio_id} / ${stream.lnb_id} / ${stream.receiver_chain_id}`, true);
        appendCell(row, point.window_index);
        appendCell(row, `[${point.start_sample}, ${point.stop_sample})`);
        appendCell(row, `[${point.start_time_s.toFixed(3)}, ${point.stop_time_s.toFixed(3)}) s`);
        appendCell(row, Number(pattern.selection_score).toFixed(6));
        appendCell(row, `${Number(pattern.winning_cfo_hz).toFixed(3)} Hz`);
        appendCell(row, `${pattern.winning_epoch_sample} samples`);
        body.append(row);
      }
    }
  }

  function draw() {
    const canvas = node("evidence-symbolwise-canvas");
    const series = [];
    for (const stream of payload?.streams || []) {
      for (let patternIndex = 0; patternIndex < 5; patternIndex += 1) {
        const first = stream.windows[0]?.patterns[patternIndex];
        if (!first || !selectedPattern(first)) continue;
        series.push({
          label: `${first.candidate_label} · ${stream.radio_id} / ${stream.lnb_id} / ${stream.receiver_chain_id}`,
          color: colors[patternIndex],
          points: stream.windows.map((window) => ({x: window.start_time_s, y: window.patterns[patternIndex].selection_score})),
        });
      }
    }
    canvas.hidden = series.length === 0;
    const legend = node("evidence-symbolwise-legend"); legend.replaceChildren();
    for (const item of series) { const span = document.createElement("span"); span.textContent = item.label; span.style.color = item.color; legend.append(span); }
    if (!series.length) return;
    const ctx = canvas.getContext("2d");
    ctx.clearRect(0, 0, canvas.width, canvas.height); ctx.strokeStyle = "#78909c"; ctx.strokeRect(56, 20, canvas.width - 76, canvas.height - 64);
    for (const item of series) {
      ctx.beginPath(); ctx.strokeStyle = item.color; ctx.lineWidth = 1.5;
      item.points.forEach((point, index) => {
        const x = 56 + (point.x / 60) * (canvas.width - 76);
        const y = 20 + (1 - Number(point.y)) * (canvas.height - 64);
        if (index === 0) ctx.moveTo(x, y); else ctx.lineTo(x, y);
      });
      ctx.stroke();
    }
    ctx.fillStyle = "#cfd8dc"; ctx.font = "14px sans-serif"; ctx.fillText("0 s", 56, canvas.height - 18); ctx.fillText("60 s", canvas.width - 62, canvas.height - 18); ctx.fillText("selection score 0–1", 62, 16);
  }

  function render() {
    const streams = payload.streams || [];
    const facts = node("evidence-symbolwise-facts"); facts.replaceChildren();
    const values = [
      ["Scope", `${streams.length} unpooled radio / authoritative LNB / RX stream${streams.length === 1 ? "" : "s"}`],
      ["Window plan", "600 exact 10 ms windows, beginning every 100 ms across 60 s"],
      ["Analyzed union", "6.000 s / 60.000 s = 10% exactly"],
      ["Semantics", "Candidate-only; calibrated detection count unavailable"],
      ["Overall derivation", payload.summary_derivation],
    ];
    for (const [label, value] of values) { const dt = document.createElement("dt"); dt.textContent = label; const dd = document.createElement("dd"); dd.textContent = value; facts.append(dt, dd); }
    const overall = node("evidence-symbolwise-overall-body"); overall.replaceChildren();
    for (const stream of streams) for (const item of stream.overall.filter(selectedPattern)) {
      const row = document.createElement("tr");
      appendCell(row, `${item.candidate_label} · ${stream.radio_id} / ${stream.lnb_id} / ${stream.receiver_chain_id}`, true);
      appendCell(row, Number(item.mean_selection_score).toFixed(6)); appendCell(row, Number(item.maximum_selection_score).toFixed(6));
      appendCell(row, `${item.winning_window_index} / ${Number(item.winning_window_start_time_s).toFixed(3)} s`);
      appendCell(row, `${Number(item.winning_cfo_hz).toFixed(3)} Hz / ${item.winning_epoch_sample} samples`); appendCell(row, item.derivation);
      overall.append(row);
    }
    node("evidence-symbolwise-window").disabled = streams.length === 0;
    draw(); renderWindow();
    setState(streams.length ? "ready" : "missing", streams.length ? `${payload.point_count} complete fixed-cadence window points loaded. Curves and summaries remain candidate-only.` : "No durable symbolwise replay stream matches the selected hardware filters.");
  }

  async function load() {
    if (!loaded) return;
    const current = ++generation;
    setState("loading", "Loading all 600 durable response points per selected hardware stream…");
    try {
      context ||= await json(`/api/v16/recordings/${encodeURIComponent(recordingId)}/evidence-context`);
      const suffix = parameters();
      if (suffix === null) {
        payload = {streams: [], point_count: 0, summary_derivation: "no hardware selected"};
        render();
        return;
      }
      const result = await json(`/api/v29/recordings/${encodeURIComponent(recordingId)}/symbolwise-replay${suffix ? `?${suffix}` : ""}`);
      if (current !== generation) return;
      if (result.candidate_only !== true || result.calibrated_detection_count !== null || result.window_count_per_stream !== 600) throw new Error("unsafe symbolwise replay semantics");
      payload = result; render();
    } catch (error) {
      if (current !== generation) return;
      payload = null; node("evidence-symbolwise-canvas").hidden = true; node("evidence-symbolwise-window-body").replaceChildren(); node("evidence-symbolwise-overall-body").replaceChildren();
      setState(error.status === 404 ? "pending" : "error", error.status === 404 ? "Durable symbolwise replay is pending or has not been published for this recording." : `Symbolwise replay failed: ${error.message}`);
    }
  }

  document.addEventListener("leo:load-extended-recording-analysis", () => { loaded = true; void load(); });
  node("evidence-controls").addEventListener("change", () => { if (loaded) void load(); });
  node("evidence-symbolwise-window").addEventListener("input", renderWindow);
})();

File: src/leo_flow/dashboard/symbolwise_replay.py
"""Contract-only projection of durable symbolwise replay for the dashboard."""

from __future__ import annotations

import math
from collections import defaultdict
from typing import Protocol

from leo_flow.contracts.dashboard_recording_evidence import (
    RecordingEvidenceContextQueryPortV0_1,
)
from leo_flow.contracts.dashboard_symbolwise_replay import (
    RecordingSymbolwiseReplayDashboardQueryV0_1,
    RecordingSymbolwiseReplayDashboardViewV0_1,
    SymbolwiseReplayDashboardStreamV0_1,
    SymbolwiseReplayPatternOverallV0_1,
    SymbolwiseReplayPatternPointV0_1,
    SymbolwiseReplayWindowPointV0_1,
)
from leo_flow.contracts.starlink_symbolwise_replay import (
    StarlinkSymbolwiseWindowEvidenceV0_1,
)
from leo_flow.contracts.starlink_symbolwise_replay_product import (
    MAXIMUM_RECORDING_REPLAY_QUERY_WINDOWS,
    RecordingStarlinkSymbolwiseReplayViewV0_1,
    StarlinkSymbolwiseReplayQueryV0_1,
    StarlinkSymbolwiseReplayStreamSelectionV0_1,
)

_SLICE_WINDOW_COUNT = 200
_OVERALL_DERIVATION = (
    "arithmetic-mean-and-maximum-selection-score-over-all-600-"
    "fixed-cadence-windows;ties-first-window"
)
_SUMMARY_DERIVATION = (
    "per-stream-per-pattern-only;arithmetic-mean-and-maximum-over-"
    "all-600-windows;no-cross-hardware-pooling"
)


class _DurableReplayQueryPort(Protocol):
    def recording_starlink_symbolwise_replay(
        self, query: StarlinkSymbolwiseReplayQueryV0_1
    ) -> RecordingStarlinkSymbolwiseReplayViewV0_1: ...


class RecordingSymbolwiseReplayDashboardProjectionV0_1:
    """Join only published contracts; storage identities never cross this seam."""

    def __init__(
        self,
        replay: _DurableReplayQueryPort,
        evidence_context: RecordingEvidenceContextQueryPortV0_1,
    ) -> None:
        self._replay = replay
        self._evidence_context = evidence_context

    def recording_symbolwise_replay_dashboard(
        self, query: RecordingSymbolwiseReplayDashboardQueryV0_1
    ) -> RecordingSymbolwiseReplayDashboardViewV0_1:
        context = self._evidence_context.recording_evidence_context(
            query.recording_id
        )
        assignments = {
            (item.radio_id, item.receiver_chain_id): item
            for item in context.receivers
            if item.recording_id == query.recording_id
            and (not query.radio_ids or item.radio_id in query.radio_ids)
            and (not query.lnb_ids or item.lnb_id in query.lnb_ids)
            and (
                not query.receiver_chain_ids
                or item.receiver_chain_id in query.receiver_chain_ids
            )
        }
        requested_receivers = tuple(
            sorted({key[1] for key in assignments})
        )
        explicitly_filtered = bool(
            query.radio_ids or query.lnb_ids or query.receiver_chain_ids
        )
        if explicitly_filtered and not assignments:
            return self._view(query, (), ())

        grouped: dict[
            tuple[str, str, str, str], list[StarlinkSymbolwiseWindowEvidenceV0_1]
        ] = defaultdict(list)
        selections: dict[
            tuple[str, str, str, str],
            StarlinkSymbolwiseReplayStreamSelectionV0_1,
        ] = {}
        reason_codes: tuple[str, ...] = ()
        for first in range(0, 600, _SLICE_WINDOW_COUNT):
            source = self._replay.recording_starlink_symbolwise_replay(
                StarlinkSymbolwiseReplayQueryV0_1(
                    recording_id=query.recording_id,
                    receiver_chain_ids=requested_receivers,
                    first_window_index=first,
                    stop_window_index=first + _SLICE_WINDOW_COUNT,
                    maximum_windows=MAXIMUM_RECORDING_REPLAY_QUERY_WINDOWS,
                )
            )
            if source.candidates_only is not True or source.truncated:
                raise RuntimeError("durable symbolwise replay slice is incomplete")
            reason_codes = source.reason_codes
            for stream in source.streams:
                selection = stream.selection
                assignment = assignments.get(
                    (selection.radio_id, selection.receiver_chain_id)
                )
                if assignment is None:
                    if explicitly_filtered:
                        continue
                    raise RuntimeError(
                        "symbolwise replay lacks authoritative LNB assignment"
                    )
                key = selection.identity
                selections[key] = selection
                grouped[key].extend(stream.windows)

        streams = tuple(
            self._stream(selections[key], assignments, tuple(grouped[key]))
            for key in sorted(grouped)
        )
        return self._view(query, streams, reason_codes)

    @staticmethod
    def _view(
        query: RecordingSymbolwiseReplayDashboardQueryV0_1,
        streams: tuple[SymbolwiseReplayDashboardStreamV0_1, ...],
        limitations: tuple[str, ...],
    ) -> RecordingSymbolwiseReplayDashboardViewV0_1:
        return RecordingSymbolwiseReplayDashboardViewV0_1(
            recording_id=query.recording_id,
            streams=streams,
            stream_count=len(streams),
            window_count_per_stream=600,
            point_count=len(streams) * 600,
            candidate_only=True,
            calibrated_detection_count=None,
            summary_derivation=_SUMMARY_DERIVATION,
            limitations=tuple(sorted(set(limitations))),
        )

    @staticmethod
    def _stream(selection, assignments, windows) -> SymbolwiseReplayDashboardStreamV0_1:
        if tuple(window.window_index for window in windows) != tuple(range(600)):
            raise RuntimeError("symbolwise replay did not provide all 600 windows")
        assignment = assignments[(selection.radio_id, selection.receiver_chain_id)]
        points = tuple(
            SymbolwiseReplayWindowPointV0_1(
                window_index=window.window_index,
                start_sample=window.start_sample,
                stop_sample=window.stop_sample,
                start_time_s=window.start_sample / selection.sample_rate_hz,
                stop_time_s=window.stop_sample / selection.sample_rate_hz,
                patterns=tuple(_pattern_point(item) for item in window.patterns),
            )
            for window in windows
        )
        overall = tuple(
            _overall(tuple(point.patterns[index] for point in points), points)
            for index in range(5)
        )
        return SymbolwiseReplayDashboardStreamV0_1(
            recording_id=assignment.recording_id,
            radio_id=selection.radio_id,
            lnb_id=assignment.lnb_id,
            receiver_chain_id=selection.receiver_chain_id,
            segment_id=selection.segment_id,
            edge=selection.edge,
            sample_rate_hz=selection.sample_rate_hz,
            frequency_center_cfo_hz=selection.frequency_center.center_cfo_hz,
            window_count=600,
            window_duration_ms=10,
            cadence_ms=100,
            analyzed_union_fraction=0.1,
            analyzed_union_percent=10.0,
            windows=points,
            overall=overall,
            candidates_only=True,
        )


def _pattern_point(item) -> SymbolwiseReplayPatternPointV0_1:
    role = item.pattern.role.value
    index = item.pattern.codebook_index
    label = (
        "Candidate-only · Qin exact"
        if index is None
        else f"Candidate-only · surrogate {index + 1}"
    )
    return SymbolwiseReplayPatternPointV0_1(
        pattern_id=item.pattern.pattern_id,
        pattern_role=role,
        codebook_index=index,
        candidate_label=label,
        selection_score=item.selection_score,
        winning_cfo_hz=item.winning_cfo_hz,
        winning_epoch_sample=item.winning_epoch_sample,
    )


def _overall(patterns, windows) -> SymbolwiseReplayPatternOverallV0_1:
    winner_index = max(
        range(len(patterns)), key=lambda index: patterns[index].selection_score
    )
    winner = patterns[winner_index]
    return SymbolwiseReplayPatternOverallV0_1(
        pattern_id=winner.pattern_id,
        pattern_role=winner.pattern_role,
        codebook_index=winner.codebook_index,
        candidate_label=winner.candidate_label,
        mean_selection_score=math.fsum(item.selection_score for item in patterns)
        / len(patterns),
        maximum_selection_score=winner.selection_score,
        winning_window_index=windows[winner_index].window_index,
        winning_window_start_time_s=windows[winner_index].start_time_s,
        winning_cfo_hz=winner.winning_cfo_hz,
        winning_epoch_sample=winner.winning_epoch_sample,
        derivation=_OVERALL_DERIVATION,
    )
